pkg/controller/persistentvolume/persistentvolume_recycler_controller.go

```diff
@@ -94,16 +116,61 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
 	return recycler, nil
 }
 
+func (recycler *PersistentVolumeRecycler) shouldRecycle(pv *api.PersistentVolume) error {
+	if pv.Spec.ClaimRef == nil {
+		return fmt.Errorf("Volume does not have a reference to claim")
+	}
+	if pv.Status.Phase == api.VolumeReleased {
+		// This is the first attempt to recycle, allow.
+		return nil
+	}
+	if pv.Status.Phase != api.VolumeFailed {
+		return fmt.Errorf("The volume is not in a final state")
+	}
+
+	// The volume is Failed, should we retry recycling?
+	retryCount := recycler.getRetryCount(pv)
+	glog.V(5).Infof("PersistentVolume[%s] %d attempts failed so far", pv.Name, retryCount)
+	if retryCount >= recycler.maximumRetryCount {
+		glog.V(5).Infof("PersistentVolume[%s] skipping deletion, %d attempts failed", pv.Name, retryCount)
+		return fmt.Errorf("%d attempts failed", retryCount)
+	}
+
+	// retry-count is fine, check the timestamp
+	lastFailureStr, ok := pv.Annotations[pvRecycleLastFailureAnnotationKey]
```

Comment:
Local cache is probably easier to handle and does not generate 'pv updated' events to other controllers, which is much better. I'm adding a patch with local cache of failed volumes (to be squashed into the final patch eventually).

On the other side, with the local cache the state is hidden inside the controller and any debugging/troubleshooting is much harder (as I found out right now). 
Annotations in `kubectl get pv -o yaml` were quite useful.